### ISSUE TITLE
Custom pluggable SerialUSB library?

### ISSUE THREAD

**Issue by elMariachi90** [Levi--G/USBLibrarySTM32#4]
First of all thank you for this amazing library which is allowing me to program my Blue Pills as game peripherals through Platformio. Can't say how much I apreciate it.

I am here to ask if it is possible in any way to add the Serial over USB functionality on top of the Game Peripheral HID.
I really need my application to send some Serial messages through USB, but I can't find any reference to the "Custom Pluggable SerialUSB library" you mentioned in your description.

Can you please help me getting this to work?

**Comment by Levi--G** [comment]
Hi, thank you for the kind words. I was about to go to bed but if I'm not mistaken there should be an example in the repo with the serial usb.
It should be along the lines of defining 
SerialUSB_ SerialUSB;
in your main and using that as your serial object. If you can't find it leave me a message and I'll respond first thing tomorrow :)

**Comment by Levi--G** [comment]
The example is here https://github.com/Levi--G/USBLibrarySTM32/blob/main/examples%2FSerialUSB%2Fsrc%2Fmain.cpp

**Comment by elMariachi90** [comment]
Wow, I didn't expect you to write back so fast.. can't thank you enough for your help!

I succeeded implementing SerialUSB + Joystick with custom HID on a chinese generic STM32F103C8 through Platformio.

Following and blending the two examples presented on this library was successfull!

**Comment by elMariachi90** [comment]
Unfortunately I found a small bug with the SerialUSB implementation.

If I plug the device into my computer, the Gaming peripheral gets recognized correctly and detects button presses with no issue. Then if i open a serial monitoring app and open the serial, it works okay and streams data back and forth with the board.. but if I close the Serial from the PC end, then the gaming peripheral gets frozen. I presume it's because the SerialUSB.print() / write functions get frozen for some reason.. indeed if I reopen the Serial monitor, the gaming peripheral resumes and works correctly.

I have tried using the SerialUSB class that you implemented to check if SerialUSB is open or not, but it seems that once is open its value keeps being true and never reverts back to false.

Can you please help with this?

**Comment by Levi--G** [comment]
Yes I will take a look tomorrow, in the mean time can you send me your code tou use, or a minimum example that would really help :)

**Comment by Levi--G** [comment]
Hi, I have taken a look at this but i see no issue, yes the serial will get flushed, but that is at usb level, your computer should still handle the cdc messages even if no program listens to the terminal port. Unless your terminal actually closes the usb device rather than the cdc port, or your code is stuck somewhere i see no reason why it would stop working when you disconnect CDC and i don't see this on my end.
So if you could send me the relevant bits of your code I can take a look, but i see no issue at the moment on my end.
The USB_Running will always return true for as long as you don't remove the usb cable that is normal, and SerialUSB should only return true when you have a terminal attached, but this can be platform/terminal dependant, it could very well be that will not return false once connected, but better like this than returning false when it is connected ;)

**Comment by numanair** [comment]
Without seeing your code it's just a guess, but might you have a 'while' statement somewhere blocking things?

**Comment by elMariachi90** [comment]
Hello guys, I didn't know it was possible to share here, but anyway here is my Platformio project.
It really seems that the SerialUSB.print hangs things up unfortunately.

Thank you guys in advance.

[Custom Handbrake Levi.zip](https://github.com/Levi--G/USBLibrarySTM32/files/14319739/Custom.Handbrake.Levi.zip)

**Comment by Levi--G** [comment]
@elMariachi90 I see what you did, can you try running without the SerialUSB.flush(); after your println. That will instruct the stm to wait until it knows the PC got the message. Does that improve?

**Comment by elMariachi90** [comment]
I have tried with and without SerialUSB.flush();
I have tried with SerialUSB.print(), SerialUSB.write()... any possible combination, but they all hang up the gaming peripheral functionality unfortunately..

I believe that any sort of SerialUSB call hangs the MCU is a frozen state when the serial port gets closed.. could you give it a try and let me know?

**Comment by Levi--G** [comment]
That is very odd, i have never seen that. ill see when i have time to check it.

**Comment by elMariachi90** [comment]
Any news?

**Comment by Levi--G** [comment]
No sorry im very busy this week, if possible could you make a minimal POC without external dependencies and how i can reproduce it? there is quite a lot of code in your project that has nothing to do with the communication, it would make it a lot easier for me. :) Also was this on a legit chip or a clone?

**Comment by Levi--G** [comment]
Ok, i have been able to reproduce the issue, however, your project is not using the standard setup, i suggest you also redo the installation of the library and follow each step closely, as your project uses the wrong defines to enable usb, which might in the long run cause different issues. :)

**Comment by Levi--G** [comment]
Version 4.0 will have configurable RX and TX buffers that will hopefully fix all freezing issues

**Comment by Levi--G** [comment]
This should be fixed in 1.0.0 release, if you still have issues you can reopen the issue.
